### PR TITLE
feat: get doc on hover from element's shortName

### PR DESCRIPTION
**Pull request by alacambre-yseop**

### CODE REVIEW

File: server/src/completion/YmlCompletionItemsProvider.ts [hunk 1]
```diff
@@ -72,4 +72,34 @@ export class YmlCompletionItemsProvider {
     public getItemByLabel(entityName: string): AbstractYmlObject {
         return this.completions.find((elem) => elem.label === entityName);
     }
+
+    /**
+     * Find *all* the completion items that has `entityName` as name.
```
Comment (mminot-yseop) [inline server/src/completion/YmlCompletionItemsProvider.ts]:
```suggestion
     * Find *all* the completion items that have `entityName` as name.
```

File: server/src/completion/YmlCompletionItemsProvider.ts [hunk 2]
```diff
@@ -72,4 +72,34 @@ export class YmlCompletionItemsProvider {
...
+     *
+     * @param entityName The name of the entity searched for.
+     * @returns The first completion item that has `entityName` as its short name, `null` otherwise.
+     */
+    public getItemByShortName(entityName: string): AbstractYmlObject {
+        return this.completions.find((elem) => elem.getShortName() === entityName);
+    }
+
+    /**
+     * Find *all* the completion items that has `entityName` as short name.
```
Comment (mminot-yseop) [inline server/src/completion/YmlCompletionItemsProvider.ts]:
```suggestion
     * Find *all* the completion items that have `entityName` as short name.
```

File: server/src/server.ts [hunk 3]
```diff
@@ -68,8 +68,10 @@ connection.onHover((_params) => {
     if (!entityName) {
         return null;
     }
-    const entity = completionProvider.getItemByLabel(entityName);
-    if (!entity || !entity.documentation) {
+    // Multiple elements can have the same shortName
+    const entities = completionProvider.getAllItemsByShortName(entityName);
+    const entity = entities.find((elem) => !!elem.documentation);
```
Comment (mminot-yseop) [inline server/src/server.ts]: What does this do? Get the first guy with nonempty doc? :thinking:

File: server/src/test/YmlCompletionItemsProvider.test.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import * as assert from 'assert';
+import { CompletionItemKind } from 'vscode-languageserver';
+
+import { YmlCompletionItemsProvider } from '../completion/YmlCompletionItemsProvider';
+import { YmlDefinitionProvider } from '../definitions';
+import { YmlLexer, YmlParser } from '../grammar';
+import { YmlKaoFileVisitor } from '../visitors';
+
+describe('YmlCompletionItemsProvider', () => {
+    it("should get be able to get a method's documentation from its short name", (done) => {
```
Comment (mminot-yseop) [inline server/src/test/YmlCompletionItemsProvider.test.ts]: Hehe

```suggestion
    it("should be able to get a method's documentation from its short name", (done) => {
```

File: server/src/test/YmlCompletionItemsProvider.test.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        const tokenStream = new CommonTokenStream(lexer);
+        const parser = new YmlParser(tokenStream);
+
+        const result = parser.kaoFile();
+        const completionProvider = new YmlCompletionItemsProvider();
+        const visitor = new YmlKaoFileVisitor(completionProvider, '', new YmlDefinitionProvider());
+        visitor.visit(result);
+        const allItemsBylabel = completionProvider.getAllItemsByLabel('getName');
+        const allItemsByShortName = completionProvider.getAllItemsByShortName('getName');
+        assert.strictEqual(allItemsBylabel.length, 1); // just the keyword getName
```
Comment (mminot-yseop) [inline server/src/test/YmlCompletionItemsProvider.test.ts]:
```suggestion
        // just the “getName” keyword
        assert.strictEqual(allItemsBylabel.length, 1);
```

File: server/src/test/YmlCompletionItemsProvider.test.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        assert.strictEqual(allItemsByShortName.length, 2); // keyword and the method
```
Comment (mminot-yseop) [inline server/src/test/YmlCompletionItemsProvider.test.ts]:
```suggestion
        // keyword and the method
        assert.strictEqual(allItemsByShortName.length, 2);
```

File: server/src/test/YmlCompletionItemsProvider.test.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        // keywords have no documentation, should get nothing
+        assert.deepStrictEqual(allItemsBylabel.filter((elem) => !!elem.documentation).length, 0);
+        // the method should be in the list; the method has the documentation.
```
Comment (mminot-yseop) [inline server/src/test/YmlCompletionItemsProvider.test.ts]: Not sure I get it.

```suggestion
        // the method should be in the list, as it has some documentation.
```

**Comment by mminot-yseop** [comment]
Test OK, I think. Waiting for quick answer to my question above and maybe a line of comment in the code if not trivial.

File: server/src/server.ts (hunk 3 above)
Comment (alacambre-yseop) [reply server/src/server.ts]: Yes, exactly.

Comment (mminot-yseop) [reply server/src/server.ts]: ### 1

Why call a “get all” method (I saw that you have equivalents with a single returned item) to eventually cut down the result to a single guy? You even lose processing power and time by scanning dudes that won’t serve any purpose. Consider, for example, giving the `(elem) => !!elem.documentation` lambda filter as a new parameter for `getAllItemsByShortName(…)` (which would become something like `getFirstItemByShortNameMatching(…)`) and stop the search as soon as you find a suitable guy.

### 2

Is it OK to pick the first? No other criterion available? :thinking:

Comment (alacambre-yseop) [reply server/src/server.ts]: = 1

Good idea with the lambda thingy. I'll do this.

= 2

No, no other criterion available yet that I may think of.

**Comment by mminot-yseop** [comment]
LOL dunno about the assertion order, but do as you wish.
Was referring to this old change:
image
You made that
> // just the “getName” keyword

part sink into oblivion and I don’t know why nor how. :woman_shrugging:
I won’t press too much for that, but still, wut.
Make sure you didn’t undo important things :open_mouth:

**Comment by alacambre-yseop** [comment]
> LOL dunno about the assertion order, but do as you wish.
> Was referring to this old change:
> image
> You made that
> 
> > // just the “getName” keyword
> 
> part sink into oblivion and I don’t know why nor how. 
> I won’t press too much for that, but still, wut.
> Make sure you didn’t undo important things

As far as I remember, I didn't test any further even before.

**Pull request merged**